src/R/Components/Impl/InteractiveWorkflow/Implementation/RSessionProviderCallback.cs

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Common.Core.Shell;
+using Microsoft.R.Components.Extensions;
+using Microsoft.R.Host.Client;
+
+namespace Microsoft.R.Components.InteractiveWorkflow.Implementation {
+    public sealed class RSessionProviderCallback : IRSessionProviderCallback {
+        private readonly ICoreShell _coreShell;
+        private readonly Lazy<IRInteractiveWorkflow> _workflowLazy;
+
+        public async Task<IntPtr> GetApplicationWindowHandleAsync() {
+            await _coreShell.SwitchToMainThreadAsync();
+            return _coreShell.ApplicationWindowHandle;
+        }
```

Comment:
It has to be async cause `GetApplicationWindowHandleAsync` may be called before VS is loaded. `SwitchToMainThreadAsync` ensures that UI thread is available so that `ApplicationWindowHandle` is correct

---
In reply to: [78041142](https://github.com/Microsoft/RTVS/pull/2268#discussion_r78041142) [](ancestors = 78041142)